Jcouls29/nethereum services

Jcouls29 opened the pull request:
I'm merging it into my other branch so that you can better see the changes. Since the other pull request never finalized, all the Unity files still show up. Once the video I put together is uploaded, I will post it here. It goes through the decisions made during the design.

Jcouls29 committed:
Implemented Nethereum into Unity through the use of .NET 4.6 class libraries.

Jcouls29 committed:
Added some more information in the README
Added some comments in the example unity project.

Jcouls29 committed:
Added ignored csproj and dlls back into the repository.

Jcouls29 committed:
Solution file was missing from source control

Jcouls29 committed:
Added Hyperbridge.Nethereum.csproj back to the repository. **Not sure why these aren't being automatically added**

Jcouls29 committed:
Fixed package references in csproj

Jcouls29 committed:
Update .gitignore to remove temp folder for unity project.

Jcouls29 committed:
Updated Wallet to allow large account searches
Added Seed Generator Interface
Added comments discussing future improvements.
Updated Documentation to include Nethereum installation
Added some tests to ensure working wallet implementation from Nethereuem

Jcouls29 committed:
Added Profile
Added Coin and Currency interfaces
Added general Ethereum Assembly so that future implementations can be used if necessary
Created basic transaction writer
Added Profile Reader/Writer
Added basic coin tests

Jcouls29 committed:
Updated namespace issue

Jcouls29 committed:
Added EtherScan to read transactions back from Ethereum BlockChain

Jcouls29 committed:
Implemented a balance read for an account

Jcouls29 committed:
Started working on services to be used by the UI.

Jcouls29 committed:
Added Basic UI Service Implementations.

Jcouls29 committed:
Fixed up issues serializing and deserializing profile

Jcouls29 committed:
Added example usage of interfaces for the UI.

Jcouls29 committed:
Removed sensitive data

Jcouls29 committed:
Removed Hyperbridge.Data assembly to simplify down by 1 assembly. Required adding Newtonsoft.Json to Core though.

Jcouls29 committed:
Reorganized folders to be clearer with intention of classes
Added missing README files
Added new method for creating a wallet that doesn't require the mnemonic phrase.

Jcouls29 committed:
Added create wallet example
Fixed bug with missing implementation of method

Jcouls29 committed:
Removed InMemoryWalletCreator as it isn't needed.

Jcouls29 committed:
Added note about assemblies.

Jcouls29 committed:
Renamed example variable to be clearer

Jcouls29 committed:
Moved files around into other projects to reduce the number of assemblies.
Renamed from Hyperbridge to Blockhub
Removed unnecessary interfaces for rsolution

Jcouls29 committed:
Reworked interfaces to simplify and be more intuitive.

Jcouls29 committed:
Renamed Namespaces from Hyperbridge to Blockhub

Jcouls29 committed:
Simplified architecture to work with concrete wallet/account classes. This was done for the sake of the Profile

Jcouls29 committed:
Fixed up some profile issues
Added Examples back in
Moved Loader/Saver to Task

Jcouls29 committed:
Fixed up namespaces

Jcouls29 committed:
Removed profile manager
Removed JsonIgnore attribute in Account

Jcouls29 committed:
Fixed up private key hiding

Jcouls29 committed:
Re-organized abstractions to make them easier to find.

Jcouls29 committed:
Re-organized a bit to better find what you're looking for.

Jcouls29 committed:
Merge remote-tracking branch 'origin/master' into jcouls29/nethereum-services

Jcouls29 commented:
[Architecture Overview](https://youtu.be/r-Sj9DAr6-g)
Here is the video discussing some of my decisions. It's a little rough (I'm a bit tired) and it also has a few audio problems here and there (so I apologize).

ericmuyser commented:
I've reviewed the video overview. Great work so far!

I like the thought put into the re-architecting and abstractions. Here's my feedback. Before we merge do you think you'll be able to verify some of these work in Unity on Windows? The obvious ones being filesystem stuff and the Nethereum calls. If that works, I will then test Mac (until we can get you one).

1. Re Importing accounts - also not sure, but initial thought is generating first 100 and scanning for balances, with an option to input specific seed indexes.

2. +1 JSON decorator pattern

3. Re - SendTransaction/ReceiveTransaction - I feel like just Transaction is fine and the user should use the FromAddress string to get the Account object if they need it

4. +1 URI as long as we get it working in Unity.

5. I feel like we should try to avoid monolithic libs and if Services and StructureMap are buckets, that suggests room for improvement. Re Github - everything should be on Github, it might just be a different repo then (perhaps this one? the Unity project itself). I understand we want to consolidate logic for the UI, but that inherently means we build a monolithic lib rather than letting the UI pick and choose what it wants with any of the various libs. I don’t know which is better. I feel like that code in Services can be put in the appropriate buckets, even if it’s higher level, it’s just additional abstraction options. I feel like it’s OK for a lib to have low level and high level exports. And for now what do you think about calling StructureMap -> UnityBridge?

6. I feel like the abstraction of ETH should be through Blockhub.Blockchain or something

7. Sorry I forgot to mention, we renamed the platform to Blockhub. Can we change the namespace?

8. I think pulling back the IEnumerable to get tokens seems fine. Could prob abstract that into a GetSupportedTokens call as well.

9. Can we rename StructureMap to UnityBridge?

10. The resolving stuff is a little cumbersome. I understand the UI shouldn't have much code complexity, but if if a dev needs to reach deeper this doesn’t prevent them in any way right?

11. Pet peeve - I don’t like the word Coin, can we use Token?

You can make these branches in this repo directly next time btw

Jcouls29 commented:
@ericmuyser So I think I understand most of what you're saying here, but I might need a few clarifications.

> 1. Re Importing accounts - also not sure, but initial thought is generating first 100 and scanning for balances, with an option to input specific seed indexes.

I will add some algorithm to include this, at least as an option. This will be a higher level API.

> 3. Re - SendTransaction/ReceiveTransaction - I feel like just Transaction is fine and the user should use the FromAddress string to get the Account object if they need it

This is perfectly fine with me! I was hesitant to begin with doing this. The thought was ease of implementation for the UI.

> 5. I feel like we should try to avoid monolithic libs and if Services and StructureMap are buckets, that suggests room for improvement. Re Github - everything should be on Github, it might just be a different repo then (perhaps this one? the Unity project itself). I understand we want to consolidate logic for the UI, but that inherently means we build a monolithic lib rather than letting the UI pick and choose what it wants with any of the various libs. I don’t know which is better. I feel like that code in Services can be put in the appropriate buckets, even if it’s higher level, it’s just additional abstraction options. 

So this is where I'm torn. I both agree and disagree with you. On the one hand, you're 100% right about the monolithic libs. I have a distaste for any assemblies that contain monolithic needs.  On the other hand, with my _limited knowledge_ of Unity, I'm not sure how we would be able to handle Dependency Injection with a more disciplined approach.

Honestly, the UI Developer can use the generics if they wish and we can get rid of the Services/Structuremap assemblies altogether. They are a higher level of abstraction, meant to make developing on this specific front-end easier. In fact, because of the extensible nature of the Desktop Client, having generics can prove to be difficult to work with. I know I've struggled before doing this myself. However, I'm completely open to suggestions on how one might implement this in Unity (and only Unity because in web apps, WPF, WinForms, there are other ways).

What I don't want to see is duplication of code on the Unity side because I simply didn't provide the necessary pieces to work with.

> I feel like it’s OK for a lib to have low level and high level exports. 

I'm in no way a programming idealist. I'm a pragmatist, so if you feel combining the Ethereum stuff together and the services into the core, is ok, I'll make an attempt to do this. The only problem is making sure the dependency graph (direction of assembly references to each other) are appropriate and don't hinder future growth.

> And for now what do you think about calling StructureMap -> UnityBridge?

No problem with this whatsoever. I'll make the change.

> 6. I feel like the abstraction of ETH should be through Blockhub.Blockchain or something

Not really sure what you're saying here exactly. Are you saying `ICoinCurrency` should be `IBlockchain` or are you saying the namespace `Hyperbridge.Core` should be `Blockhub.Blockchain`? I'm fine with either or both.

> 7. Sorry I forgot to mention, we renamed the platform to Blockhub. Can we change the namespace?

Yep. Sure thing. 👍 

> 8. I think pulling back the IEnumerable to get tokens seems fine. Could prob abstract that into a GetSupportedTokens call as well.

This is a great idea. Actually I might consider how I can use this to better leverage a higher level API. Maybe let this be the start of a redesign of those (pending the discussion of **5.**).

> 9. Can we rename StructureMap to UnityBridge?

Will do. It's more clear that way anyways.

> 10. The resolving stuff is a little cumbersome. I understand the UI shouldn't have much code complexity, 

Goes back to the discussion in **5.**. How do we handle DI? Inversion of Control in Unity? Are we just going to attach everything we need into a single MonoBehaviour and pass that around? I'm ok with this, I just need some guidance on the Unity side.

> but if if a dev needs to reach deeper this doesn’t prevent them in any way right?

Of course. Again, we can get rid of Services/StructureMap altogether and still everything will work appropriately.

> 11. Pet peeve - I don’t like the word Coin, can we use Token?

Works for me. I'll make the change.

> You can make these branches in this repo directly next time btw

Ok. For this pull request, I'll finish it here in mine since I need to keep it updated, but the next round, I'll do on the hyperbridge repo.

Jcouls29 commented:
> Before we merge do you think you'll be able to verify some of these work in Unity on Windows?

I will put together another separate Unity project (like before) that runs through these functions to ensure they are working properly with Unity.

Jcouls29 commented:
@ericmuyser @lopezmramon 

[Zenject (Unity IoC)](https://github.com/modesttree/Zenject)

I saw this and think it might be a viable solution for working with DI in Unity3D

ericmuyser commented:
> Not really sure what you're saying here exactly. Are you saying ICoinCurrency should be IBlockchain or are you saying the namespace Hyperbridge.Core should be Blockhub.Blockchain? I'm fine with either or both.

I think Hyperbridge.Services should be Blockhub.Blockchain. Hyperbridge.StructureMap would be Blockchain.UnityBridge. Thoughts?

> Goes back to the discussion in 5.. How do we handle DI? Inversion of Control in Unity? Are we just going to attach everything we need into a single MonoBehaviour and pass that around? I'm ok with this, I just need some guidance on the Unity side.

Zenject looks great. I think using a singleton MonoBehaviour that do the binding in the `Awake` method, while other components use `Start` instead, would work. (Awake being called before Start). I think it'll require some testing to know the best way.

> I will put together another separate Unity project (like before) that runs through these functions to ensure they are working properly with Unity.

Great, that could be a subfolder ala `Examples/Unity`

Jcouls29 commented:
> I'm not saying Ethereum stuff should be in Blockhub.Core, but that whatever lib is responsible for Ethereum commun can have low and high level stuff. More on that below.

So I have combined Ethereum stuff in the Ethereum lib and moved the other interfaces to the Core. We were on the same page here. Sometimes combining low/high level can be hard to wrap my head around. In all reality, I have kind of pushed them all into the same level.

> I think Hyperbridge.Services should be Blockhub.Blockchain. Hyperbridge.StructureMap would be Blockchain.UnityBridge. Thoughts?

I have actually removed both. Services has been combined to other assemblies. StructureMap should be handled in the UI with Zenject (imho). If you disagree, please let me know.

I'm still working on the Unity integration. Will be a basic example of everything. A lot has changed to try and simplify. Please take another look @ericmuyser

Jcouls29 committed:
Renamed folders to Blockhub

Jcouls29 committed:
Fixed .gitignore for the C# solution

Jcouls29 committed:
fixed untracked files

Jcouls29 commented:
@ericmuyser I had to fix the gitignore so things could properly be committed.  I shouldn't be making any other commits here until the Unity portion is complete. Any other thoughts, let me know.

Jcouls29 committed:
Added more appropriate Exceptions to services

Jcouls29 committed:
Added Ambient Context Pattern for Profile
Added Decorators to support auto-saving of profile when creating wallets and accounts

Jcouls29 committed:
Renamed some interfaces to be consistent
Added profile loader that sets the context on-load

Jcouls29 committed:
Renamed Bip39SeedGenerator to be consistent with the rest of the framework

Jcouls29 committed:
Added .editorconfig to force LF line endings and SPACE instead of TAB

Jcouls29 committed:
Added .gitattributes to control line endings

Jcouls29 committed:
Added Unity Project for Nethereum Services and completed Zenject Dependency Installer

Jcouls29 committed:
Finished Unity Example Application
Switched out System.Numerics for NBitcoin.
Fixed some bugs.

Jcouls29 commented:
@ericmuyser Added the Unity Example Application. Had some issues that I wasn't expecting.
I'm ill-experienced working with numbers as large as we're dealing with in Ethereum. I had to convert from System.Numerics.BigInteger to the NBitcoin version (which is lacking in features). This pointed out some interesting issues with dealing with these numbers. We may have to re-consider how to deal with these large numbers (I currently don't have any ideas). 

Please let me know your thoughts.

Jcouls29 committed:
Renamed Token to Currency and TokenSource to BlockchainType

Jcouls29 committed:
Fixed WEI to ETHER and ETHER to WEI conversions to be arbitrarily large.

Jcouls29 committed:
Tested Unity Example

Jcouls29 committed:
updated gitignore

Jcouls29 committed:
Added DLLs to examples to ensure proper running on other systems without requiring a build.

Jcouls29 commented:
@ericmuyser I have finished updating the currency conversions to handle arbitrarily large numbers past the 9.22 64-bit limit. It is still using NBitcoin as there were problems adding the types to our system (being unsafe assemblies).

At this point, it is working on Windows. Please review and provide feedback.

ericmuyser commented:
Really loving this Justin, great work so far.

I noticed a lot of the Blockhub.Core stuff isnt in the Blockhub.Core namespace, its either in just Blockhub or something like Blockhub.Wallet. Is that common practice? I felt like the name of the package and the namespace should be consistent, no? Should Blockhub.Wallet be its own thing? And the rest use Blockhub.Core rather than just Blockhub?

Architecturally we prefer to use templates AND interfaces? Rather than just using interfaces and casting if necessary (rarely)? Just want to make this point explicit/have documentation around the reasoning

```
public class AutoSaveProfileAccountCreate<T> : IAccountCreate<T> where T : IBlockchainType
```

That could have just been `public class AutoSaveProfileAccountCreate : IAccountCreate` and the methods use IBlockchainType rather than T. Which is what you usually see, and usually works fine unless you need something specific to the derived class, which would require casting. Is that right? Sorry I’m a bit rusty about these conventions in web land for a while.

I'm thinking we could use more tests around the wallet/profile stuff. Thoughts?

Jcouls29 commented:
@ericmuyser Thanks Eric for going through this. Here are my responses:

> I noticed a lot of the Blockhub.Core stuff isnt in the Blockhub.Core namespace, its either in just Blockhub or something like Blockhub.Wallet. Is that common practice? I felt like the name of the package and the namespace should be consistent, no? 

Man gotta dig deep here. This is how Microsoft does some of their assembly naming. The project will build to Blockhub.Core.dll but this is meant to represent the core interfaces and assembly everything builds off of right? For example, there is a 'System.Core.dll' from Microsoft as we all 'Microsoft.Build.Tasks.Core.dll', etc. etc. I'm not married to this. It can be Blockhub.dll. But the core usually implies that it is not a monolithic assembly. Does that make sense?

> Should Blockhub.Wallet be its own thing? And the rest use Blockhub.Core rather than just Blockhub?

This is where your business rules knowledge will help. If you feel that the `Wallet` will become the central focus of the application and there will be a lot of code wrapped around it in the future, then that makes sense to me. But I don't see the big picture yet (too early for me to understand it). With that said, Wallet is extensively used in the system now, so just let me know what you think is best here.

> Architecturally we prefer to use templates AND interfaces? Rather than just using interfaces and casting if necessary (rarely)? Just want to make this point explicit/have documentation around the reasoning

```public class AutoSaveProfileAccountCreate<T> : IAccountCreate<T> where T : IBlockchainType```

> That could have just been public class AutoSaveProfileAccountCreate : IAccountCreate and the methods use IBlockchainType rather than T. Which is what you usually see, and usually works fine unless you need something specific to the derived class, which would require casting. Is that right? Sorry I’m a bit rusty about these conventions in web land for a while.

So in the .NET world I believe templates are called generics, right? You're referring to the `T` type parameter? Well you're right. I could just as easily utilize `IAccountCreate` instead, however I will have to propagate that through to everything. So everything will be `IAccountCreate` and `IAccountCreate<T>` would be removed. The particular class you reference above is utilizing the [decorator pattern](https://en.wikipedia.org/wiki/Decorator_pattern) to extend functionality "dynamically" to the base objects. So I think the bigger question might be why I decided to utilize `IAccountCreate<T>` over `IAccountCreate` and then just have that passed into other interfaces.

My reasoning is simple, I don't want BitCoin Accounts/Wallets passed to Ethereum Classes.  Then explicit casting is required. This is weak-typing and whenever possible, I like things to be strong-typed.

Supplemental Video: [Reasoning for Generics](https://youtu.be/gTaj0mbx3rU)

> I'm thinking we could use more tests around the wallet/profile stuff. Thoughts?

Sure. Are you guys a "Tests act as the UI" shop or "Tests should be on every class" shop? This will change how I approach the tests. I lean towards "Tests act as the UI" because tests will only be testing the business rules and not be coupled to the architecture. But if you prefer a 1 to 1 test class to concrete class, I can do that as well.

ericmuyser commented:
Sorry the above comments came in late - forgot to finish the review

ericmuyser commented:
> Man gotta dig deep here. This is how Microsoft does some of their assembly naming. The project will build to Blockhub.Core.dll but this is meant to represent the core interfaces and assembly everything builds off of right? For example, there is a 'System.Core.dll' from Microsoft as we all 'Microsoft.Build.Tasks.Core.dll', etc. etc. I'm not married to this. It can be Blockhub.dll. But the core usually implies that it is not a monolithic assembly. Does that make sense?

I'm okay with Blockhub.Core, I'm just not a fan of bundling namespaces that arent in that in there. It sounds like this is a bucket of things we don't have a bucket for... and there's probably a better name for that since those things may not exactly be Core to the app. (Yes you'll need my business logic on that front). Yah I think Blockhub.dll is better for now until we sort that out. I think we should also have a Blockhub.Core.dll for things in the Blockhub.Core namespace, and a Blockhub.Wallet.dll for those.. you get the idea.

> So in the .NET world I believe templates are called generics, right? You're referring to the T type parameter? Well you're right.

Right and right.

> My reasoning is simple, I don't want BitCoin Accounts/Wallets passed to Ethereum Classes. Then explicit casting is required. This is weak-typing and whenever possible, I like things to be strong-typed.

Makes sense. Ah yes strong typing, coming back to me. Usually see the other option you explained in the vid (dynamic checks). Looks good to me.

ericmuyser commented on Framework/Blockhub.Core/Services/Wallet/SearchForAccountsByBalanceWalletCreate.cs:
```diff
@@ -0,0 +1,65 @@
+            return await SearchAndFillInAccounts(await WalletCreator.CreateWallet(name, password));
+        }
+
+        private async Task<Wallet<T>> SearchAndFillInAccounts(Wallet<T> wallet)
```
Seems a little strange that this class takes a wallet and returns the same wallet, modifying the state within. Do you think it would be better if this was `SearchForAccounts` and returned the list of accounts? The user can then add if they want, or maybe they just want to check them.

Jcouls29 replied:
I will change this for simplicity. The other way is better when we want to add functionality without changing code in the UI. It auto-fills in the accounts and follows the open/closed principle 

[Uncle Bob](https://8thlight.com/blog/uncle-bob/2014/05/12/TheOpenClosedPrinciple.html)

[Example](http://joelabrahamsson.com/a-simple-example-of-the-openclosed-principle/)

I used decoration to perform the open/closed principle here.

ericmuyser commented on Framework/Blockhub.Core/Services/Wallet/SearchForAccountsByBalanceWalletCreate.cs:
```diff
@@ -0,0 +1,65 @@
+        private IBalanceRead<T> BalanceReader { get; }
+        private int SearchCount { get; }
+
+        public SearchForAccountsByBalanceWalletCreate(IWalletCreate<T> walletCreator, IAccountCreate<T> walletManager, IBalanceRead<T> balanceReader) :
```
Could we do wallet creation and searching for accounts separately? There's a lot of code here for something that seems like it could be a couple helper methods. The class name is a little unclear too, if I had to rename maybe `SearchForAccountsByBalanceAndCreateWallet` or `AccountByBalanceWalletCreator` but I would wonder why I don't just have code like this:

```
Wallet wallet = new Wallet(...);
Accounts[] accounts = Wallet.SearchForAccounts(wallet);
```

Jcouls29 replied:
Ok. I will put something together for this.

Jcouls29 commented on Examples/Nethereum Services/Assets/DependencyInstaller.cs:
```diff
@@ -0,0 +1,132 @@
+using Blockhub.Nethereum;
+using Blockhub.Ethereum;
+using Blockhub.Transaction;
+using Blockhub.EtherScan;
```
I will change this.

Jcouls29 commented on Framework/Blockhub.Ethereum/Models/Ether.cs:
```diff
@@ -0,0 +1,119 @@
+
+namespace Blockhub.Ethereum
+{
+    public class Ether : ICurrency<Ethereum>
```
what does this mean?

ericmuyser replied:
Awesome class 👍

Jcouls29 commented on Framework/Blockhub.Core/Models/Transaction/Transaction.cs:
```diff
@@ -0,0 +1,35 @@
+using Newtonsoft.Json;
+using System;
+
+namespace Blockhub.Services
```
I will rename this appropriately

Jcouls29 commented on Framework/Blockhub.Core/Services/Wallet/SearchForAccountsByBalanceWalletCreate.cs:
```diff
@@ -0,0 +1,65 @@
+                    // Let's make sure this account doesn't exist for any reason in the wallet already
+                    if (!wallet.Accounts.Any(x => x.Address.Equals(account.Address, StringComparison.OrdinalIgnoreCase)))
+                    {
+                        foundCount++;
```
This was being used in the Examples. Will be removing into a static method.

Jcouls29 commented on Framework/Blockhub.Nethereum/Services/Bip39SeedGenerate.cs:
```diff
@@ -0,0 +1,33 @@
+    /// <summary>
+    /// Generates a Mnemonic Phrase indepedently of the wallet implementation.
+    /// </summary>
+    public class Bip39SeedGenerate : ISeedGenerate<string>
```
I can. It uses NBitcoin and since I moved that into the core library, we can move it now. Will do

Jcouls29 committed:
Updated Namespaces and added IAccountSearcher according to PR4
Added Tests and Emulated Services
Updated Unity Example Project

ericmuyser commented:
This looks good to me. I think we're ready to rock n roll. I'll give Ramon a chance to review.

ericmuyser commented:
All green 🚀 🚀 🚀
